File: Workshop1_Introduction/Warmup_Fizz Buzz_412.py
# UCLA ACM ICPC: Interview Track Leetcode Problem Solutions

# Disclaimer: This is not the only valid solution and we do not claim our solutions 
# to be optimal in terms of runtime or memory usage, we are merely giving example 
# solutions to questions for learning purposes only

# Fizz Buzz
# Leetcode Problem 412
# https://leetcode.com/problems/fizz-buzz/

class Solution(object):
    def fizzBuzz(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        # initialize an empty array to store the output we wish to return
        ret = []
        
        # we want our values outputted to start at 1 and go up to n
        # Python range(x,y) values include x, and go up to (but don't include) y
        for i in range(1, n + 1):
        
        # Note: we know mathematically that if a % b == 0, then a is divisible by b

            # if our current value i is divisible by both 3 and 5, add "FizzBuzz" to the output
            if i % 3 == 0 and i % 5 == 0:
                ret.append("FizzBuzz")
            
            # otherwise, if our current value i is divisible by 3, add "Fizz" to the output
            elif i % 3 == 0:
                ret.append("Fizz")
            
            # otherwise, if our current value i is divisible by 5, add "Buzz" to the output
            elif i % 5 == 0:
                ret.append("Buzz")
            
            # at this point we know i is not divisible by 3, 5, or both, so we add the integer (as a string) to the output
            else:
                ret.append(str(i))
        
        # return our array of outputs
        return ret

File: Workshop1_Introduction/Warmup_Fizz Buzz_412.cpp
/*
    UCLA ACM ICPC: Interview Track Leetcode Problem Solutions

    Disclaimer: This is not the only valid solution and we do not claim our solutions 
    to be optimal in terms of runtime or memory usage, we are merely giving example 
    solutions to questions for learning purposes only

    Fizz Buzz
    Leetcode Problem 412
    https://leetcode.com/problems/fizz-buzz/
*/

class Solution {
public:
    // Note: Vectors represent arrays that can change in size.
    // To use vectors, you need to include the vector library,
    // which Leetcode already does for you
    vector<string> fizzBuzz(int n) {
        // Initialize an empty vector to store the output we wish to return
        vector<string> result;

        // We want our vector to represent numbers from 1 to n (inclusive)
        for(int i = 1; i <= n; i++) {
            // Note: We know that if a % b == 0, then a is divisible by b

            // If our current value i is divisible by both 3 and 5,
                // Add "FizzBuzz" to the vector
            if(i % 3 == 0 && i % 5 == 0)
                // Note: vector::push_back("FizzBuzz") adds a new element
                // at the end of the vector, after its current last element.
                // "FizzBuzz" is copied (or moved) to the new element
                result.push_back("FizzBuzz");

            // Otherwise, if our current value i is divisible by 3,
                // Add "Fizz" to the output
            else if(i % 3 == 0)
                result.push_back("Fizz");

            // Otherwise, if our current value i is divisible by 5,
                // Add "Buzz" to the output
            else if(i % 5 == 0)
                result.push_back("Buzz");

            // At this point we know i is not divisible by 3, 5, or both,
                // So we add the integer (as a string) to the output
            else
                // to_string(i) returns a string with the representation of i
                // Note: To use it, you need to include the string library,
                // which Leetcode already does for you
                result.push_back(to_string(i));
        }
        return result;
    }
};
